### PR TITLE
Presearch Code Refactor

### PR DESCRIPTION
**Pull request by CascadingRadium**
- Refactor the presearch code path to make it more generic and extensible.
- Add an ExtractFields API for obtaining the set of fields applicable to a generic query.
- Add support for a new API to set the index mapping for an alias.
  This is useful when an alias contains partitions of the same index, as the index 
  mapping would be consistent across all indexes and can be inferred directly from 
  the alias.

### CODE REVIEW

**Comment by Thejas-bhat** [comment]
my bad, rebased (to the latest master) but pushed on the wrong branch :p

**Comment by Thejas-bhat** [comment]
@CascadingRadium, so I looked into the bm25 code and realized that we are going to aggregate a field specific stat as well. 

in your synonyms code, i see you've written a query tree parser

```
// ExtractSynonyms extracts synonyms from the query tree and returns a map of
// field-term pairs to their synonyms. The input query tree is traversed and
// for each term query, the synonyms are extracted from the synonym source
// associated with the field. The synonyms are then added to the provided map.
// The map is returned and may be nil if no synonyms were found.
```

can you refactor that to two sets of APIs one generic one which returns all the fields in the query and synonyms can use that API in ExtractSynonyms and bm25 can also use this API to fetch those field specific stat

File: index_alias_impl.go
Comment (Thejas-bhat) [inline index_alias_impl.go:192]: i dont think we should do this check here, since redistributePreSearchData() internally is supposed to redistribute knn/bm25/synonym specific pre search data based on the key set in req object.

**Comment by abhinavdangeti** [comment]
Put a `do not merge` label here as this is already included within https://github.com/blevesearch/bleve/pull/2090 .

**Comment by abhinavdangeti** [comment]
-or- if you think we merge this first and rebase #2090 that's fine too.

**Comment by CascadingRadium** [comment]
i think we can merge this and rebase the https://github.com/blevesearch/bleve/pull/2090 on top

File: search.go
Comment (CascadingRadium) [reply search.go:593]: currently i am using the API in `search_knn.go` in this PR, but this becomes really helpful in the synonyms code where it gets used across 3 - 4 files and as `search_knn.go` is under a build tag, i have put it in `search.go`

File: index_alias_impl.go
Comment (CascadingRadium) [reply index_alias_impl.go:627]: I don't understand... the `preSearchRequired` returns either nil (if no presearch required) or a pointer to a struct... which helps to discern whether a presearch is required.

```
	flags, err := preSearchRequired(req, i.mapping)
	if err != nil {
		return nil, err
	}
	if req.PreSearchData == nil && flags != nil {

```

Comment (abhinavdangeti) [reply index_alias_impl.go:627]: I'm talking about the function parameters, `preSearchFlags` is a struct with a single bool - where I assume you'll add more in the future. This doesn't have to be a function pointer. Recommend changing the function signature here and everywhere else it's used to ..

```
func requestSatisfiedByPreSearch(req *SearchRequest, flags preSearchFlags) bool {
```

Comment (abhinavdangeti) [reply index_alias_impl.go:627]: You won't need to do that `flags != nil` check anywhere then.

Comment (CascadingRadium) [reply index_alias_impl.go:627]: hmm ok, flags == nil was used to indicate that the presearch is not required, but whit this change, an object will be created for every query, and the preSearchRequired method becomes below, is that fine ?

```

func preSearchRequired(req *SearchRequest, m mapping.IndexMapping) (preSearchFlags, bool) {
	// Check for KNN query
	knn := requestHasKNN(req)
	if knn {
		return preSearchFlags{
			knn: knn,
		}, true
	}
	return preSearchFlags{}, false
}

```

Comment (abhinavdangeti) [reply index_alias_impl.go:627]: I see, so if we add more flags into this construct, your point is we'll need to check for all of them to be false to skip presearch - if that's the case, I'm fine with leaving this as is where you make sure to check flags == nil.

Comment (CascadingRadium) [reply index_alias_impl.go:628]: ok done

**Review by abhinavdangeti — APPROVED** [review]
Wait for one more review perhaps. Also @CascadingRadium hope you've made sure all the knn tests still succeed with this change.

**Review by Likith101 — APPROVED** [review]

**Pull request merged**
